envoy/server.go

```diff
@@ -0,0 +1,124 @@
+package envoy
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Nitro/sidecar/catalog"
+	"github.com/Nitro/sidecar/config"
+	"github.com/Nitro/sidecar/envoy/adapter"
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	xds "github.com/envoyproxy/go-control-plane/pkg/server"
+	"github.com/relistan/go-director"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+type xdsCallbacks struct{}
+
+func (*xdsCallbacks) OnStreamOpen(context.Context, int64, string) error  { return nil }
+func (*xdsCallbacks) OnStreamClosed(int64)                               {}
+func (*xdsCallbacks) OnStreamRequest(int64, *api.DiscoveryRequest) error { return nil }
+func (*xdsCallbacks) OnStreamResponse(_ int64, req *api.DiscoveryRequest, _ *api.DiscoveryResponse) {
+	if req.GetErrorDetail().GetCode() != 0 {
+		log.Errorf("Received Envoy error code %d: %s",
+			req.GetErrorDetail().GetCode(),
+			strings.ReplaceAll(req.GetErrorDetail().GetMessage(), "\n", ""),
+		)
+	}
+}
+func (*xdsCallbacks) OnFetchRequest(context.Context, *api.DiscoveryRequest) error   { return nil }
+func (*xdsCallbacks) OnFetchResponse(*api.DiscoveryRequest, *api.DiscoveryResponse) {}
+
+// Server is a wrapper around Envoy's control plane xDS gRPC server and it uses
+// the Aggregated Discovery Service (ADS) mechanism.
+type Server struct {
+	Listener      catalog.Listener
+	config        config.EnvoyConfig
+	state         *catalog.ServicesState
+	snapshotCache cache.SnapshotCache
+	xdsServer     xds.Server
+}
+
+// Run sets up the Sidecar listener event loop and starts the Envoy gRPC server
+func (s *Server) Run(ctx context.Context, looper director.Looper, grpcListener net.Listener) {
+	grpcServer := grpc.NewServer()
+	envoy_discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, s.xdsServer)
+
+	go looper.Loop(func() error {
+		// Block until we get an event indicating a state change.
+		// We discard the event since we need a snapshot of the entire state.
+		<-s.Listener.Chan()
+
+		// When a server is expired in catalog/services_state.go -> ExpireServer(),
+		// the listener will receive an event for each expired service. We want to
+		// flush the channel to prevent rapid-fire updates to Envoy.
+		// This was inspired from receiver/receiver.go -> ProcessUpdates().
+		// TODO: Think of a more aggressive / reliable way of draining since we
+		// used a larger value for listenerEventBufferSize.
+		pendingEventCount := len(s.Listener.Chan())
+		for i := 0; i < pendingEventCount; i++ {
+			<-s.Listener.Chan()
+		}
+
+		listeners, err := adapter.EnvoyListenersFromState(s.state, s.config.BindIP)
```

Comment:
Both `adapter.EnvoyListenersFromState()` and `adapter.EnvoyClustersFromState()` lock state. I could move the locking here, so we lock only once.

Reply:
Shorter-lived locks are better. The longer we hold a lock, the worse it is for interleaving updates from gossip while we do this.

Reply:
Indeed, I'm overthinking it in terms of performance. Will leave it as is.